README.md
TibcoEMSMonitor
===============

## Introduction

An AppDynamics Machine Agent add-on to report metrics from a [Tibco EMS
Server][] and its queues.

Tibco EMS is messaging middleware that provides persistent queues as well as
a publish/subscribe mechanism. It can be used as a JMS provider, or it can be
used directly via native API's.

This extension requires the Java Machine Agent.


## Prerequisites

Before starting this monitor, make sure your EMS server is configured to report
statistics. You can do this by editing the file `tibemsd.conf` in your
`TIBCO_HOME`, or by using the `tibemsadmin` command line utility.


#### Configuring statistics by editing tibemsd.conf

1. Add the following line to `tibemsd.conf`:

        statistics = enabled

1. Restart Tibco EMS.


#### Configuring statistics using the command line

Use the `tibemsadmin` utility to change the server configuration.

        root# tibemsadmin -server localhost:7222

        TIBCO Enterprise Message Service Administration Tool.
        Copyright 2003-2013 by TIBCO Software Inc.
        All rights reserved.

        Version 8.0.0 V9 6/7/2013

        Login name (admin):
        Password:
        Connected to: tcp://localhost:7222
        Type 'help' for commands help, 'exit' to exit:
        tcp://localhost:7222> set server statistics=enabled
        Server parameters have been changed
        tcp://localhost:7222> quit
        bye


## Installation

1. Download TibcoEMSMonitor.zip from the Community.
2. Copy TibcoEMSMonitor.zip into the directory where you installed the
   machine agent, under `$AGENT_HOME/monitors`.
3. Unzip the file. This will create a new directory called `TibcoEMSMonitor`.
4. In `$AGENT_HOME/monitors/TibcoEMSMonitor`, edit the file `monitor.xml` and
   configure the extension for your Tibco EMS installation.
5. Restart the machine agent.


## Configuration

Configuration for this monitor is in the `monitor.xml` file in the monitor
directory. All of the configurable options are in the `<task-arguments>`
section.

hostname
: Name or IP address of the Tibco EMS server. Required.

port
: TCP port number where the Tibco server is listening. The default value is
  7222. Required.

userid
: Administrative user ID for the Tibco admin interface. The default value is
  "admin". Required.

password
: Password for the administrative user ID. The default value is an empty
  password. Required.

tier
: Name of the tier in AppDynamics for which the monitor should register its
  metrics. If not specified, the metrics will be registered on all tiers.
  Optional.

emsservername
: An additional folder to create under the "Custom Metrics" folder.
  Optional.

showTempQueues
: If set to true, the monitor will report metrics on temporary queues (defined
  as any queue whose name starts with `$TMP$.`). **NOTE:** Enabling this option
  can potentially cause the agent to overflow its metric limit.

showSysQueues
: If set to true, the monitor will report metrics on system queues (defined as
  any queue whose name starts with `$sys.`).




## Metrics Provided

### Global Instance Metrics

| Metric Name          | Description                                                           |
| :------------------- | :-------------------------------------------------------------------- |
| DiskReadRate         | Rate at which messages are being read from disk, in bytes per second  |
| DiskWriteRate        | Rate at which messages are being written to disk, in bytes per second |
| ConnectionCount      | Current number of connections                                         |
| MaxConnections       | Maximum number of connections allowed by the server                   |
| ProducerCount        | Total number of producers                                             |
| ConsumerCount        | Total number of consumers                                             |
| PendingMessageCount  | Total number of pending messages                                      |
| PendingMessageSize   | Total size of pending messages in bytes                               |
| InboundMessageCount  | Number of inbound messages                                            |
| InboundMessageRate   | Number of inbound messages per second                                 |
| InboundBytesRate     | Volume of inbound bytes per second                                    |
| OutboundMessageCount | Number of outbound messages                                           |
| OutboundMessageRate  | Number of outbound messages per second                                |
| OutboundBytesRate    | Volume of outbound bytes per second                                   |


### Per-Queue Metrics

| Metric Name           | Description |
| :-------------------- | :---------- |
| ConsumerCount         | Number of consumers for this destination |
| ReceiverCount         | Number of active receivers on this queue |
| DeliveredMessageCount | Total number of messages that have been delivered to consumer applications but have not yet been acknowledged |
| PendingMessageCount   | Total number of pending messages for this destination |
| InTransitCount        | Total number of messages that have been delivered to the queue owner but have not yet been acknowledged |
| FlowControlMaxBytes   | Volume of pending messages (in bytes) at which flow control is enabled for this destination |
| PendingMessageSize    | Total size for all pending messages for this destination |
| MaxMsgs               | Maximum number of messages that the server will store for pending messages bound for this destination |
| MaxBytes              | Maximum number of message bytes that the server will store for pending messages bound for this destination |
| InboundByteRate       | Bytes received per second |
| InboundMessageRate    | Messages received per second |
| InboundByteCount      | Total number of bytes received |
| InboundMessageCount   | Total number of messages received |
| OutboundByteRate      | Bytes sent per second |
| OutboundMessageRate   | Messages sent per second |
| OutboundByteCount     | Total number of bytes sent |
| OutboundMessageCount  | Total number of messages sent |


## Caution

This monitor can potentially register hundred of new metrics, depending on how
many queues are in EMS. By default, the Machine Agent will only report 200
metrics to the controller, so you may need to increase that limit when
installing this monitor. To increase the metric limit, you must add a parameter
when starting the Machine Agent, like this:

    java -Dappdynamics.agent.maxMetrics=1000 -jar machineagent.jar

Please note that the maximum value you can provide is 5000.


## Support

For any questions or feature requests, please contact the [AppDynamics Center
of Excellence][].

**Version:** 2.2.0  
**Controller Compatibility:** 3.6 or later  
**Last Updated:** 11/17/2013  
**Author:** Todd Radel  

------------------------------------------------------------------------------

## Release Notes

### Version 2.2.0
- Cleaned up directory structure.
- Rebuilt Ant scripts.

[Tibco EMS server]: http://www.tibco.com/products/automation/messaging/enterprise-messaging/enterprise-message-service/default.jsp
[AppDynamics Center of Excellence]: mailto:ace-request@appdynamics.com
[help@appdynamics.com]: mailto:help@appdynamics.com


CHANGES.md
### Version 2.2.0
- Cleaned up directory structure.
- Rebuilt Ant scripts.

### Version 2.1.1
- Added inbound and outbound message/byte count and rate for each queue.
- General code cleanup.
- Re-released to AppSphere.

### Version 1.0.0
* Initial open source release 


src/test/java/com/appdynamics/extensions/tibcoems/TibcoEMSMonitorTest.java
package com.appdynamics.extensions.tibcoems;

import com.tibco.tibjms.admin.ConnectionInfo;
import com.tibco.tibjms.admin.ConsumerInfo;
import com.tibco.tibjms.admin.QueueInfo;
import com.tibco.tibjms.admin.TibjmsAdmin;
import com.tibco.tibjms.admin.TibjmsAdminException;


public class TibcoEMSMonitorTest {

	String[] serverURL = {"tcp://localhost:7222"};
	String userid = "admin";
	String password = "";
	
	public void start() {
		try {
			TibjmsAdmin tibcoAdmin = new TibjmsAdmin(serverURL[0], userid, password);
			
			System.out.println("Connection Information");
			ConnectionInfo[] connectionInformation =  tibcoAdmin.getSystemConnections();
			for (int i = 0; i < connectionInformation.length; i++) {
				ConnectionInfo connectionInfo = connectionInformation[i];
				System.out.println("=====================================");
				String host = connectionInfo.getHost();
				System.out.println("Host\t"+ host);
				System.out.println("Connection Product Count\t" +  connectionInfo.getProducerCount());
				System.out.println("Address\t" + connectionInfo.getAddress());
				System.out.println("Client Type\t" + connectionInfo.getClientType());
				System.out.println("Session Count\t " + connectionInfo.getSessionCount());
			}
			
			
			ConsumerInfo[] consumers = 	 tibcoAdmin.getConsumersStatistics();
			System.out.println("Consumer Information size="+consumers);
			
			for (int i = 0; i < consumers.length; i++) {
				ConsumerInfo consumerInfo = consumers[i];
			    System.out.println("Durable Name\t\t" + consumerInfo.getDurableName());
			    System.out.println("===============================");
			    
			    
			    ConsumerInfo.Details details = consumerInfo.getDetails();
			    if (details != null) {
			    	System.out.println("ElapsedSinceLastSent " + consumerInfo.getDetails().getElapsedSinceLastSent());
			    	System.out.println("TotalAcknowledgedCount " + consumerInfo.getDetails().getTotalAcknowledgedCount());
			    	System.out.println("CurrentMsgCountSentByServer " + consumerInfo.getDetails().getCurrentMsgCountSentByServer());
			    }
			    System.out.println("DestinationName\t\t" + consumerInfo.getDestinationName());
			    System.out.println("Username\t\t" + consumerInfo.getUsername());
			    System.out.println("PendingMessageCount\t\t" + consumerInfo.getPendingMessageCount());
			    System.out.println("PendingMessageSize\t\t" + consumerInfo.getPendingMessageSize());
			    System.out.println("Selector\t\t" + consumerInfo.getSelector());
				System.out.println("ConnectionID\t" + consumerInfo.getConnectionID());
				System.out.println("");
			}
			
			System.out.println("");
			System.out.println("Queue Information");
			QueueInfo[] queueInformation = tibcoAdmin.getQueuesStatistics();
			for (int i = 0; i < queueInformation.length; i++) {
				QueueInfo queueInfo = queueInformation[i];
				System.out.println(queueInfo.getName());
				System.out.println("================================");
				System.out.println("Consumer Count\t\t\t" + queueInfo.getConsumerCount());
				System.out.println("Delivered Message Count\t\t" + queueInfo.getDeliveredMessageCount());
				System.out.println("Flow Control Max Bytes\t\t" + queueInfo.getFlowControlMaxBytes());
				System.out.println("Pending Message Count\t\t" + queueInfo.getPendingMessageCount());
				System.out.println("");
			}
			
			
			System.out.println("Closing Connection to Server");
			tibcoAdmin.close();
			
		
		} catch (TibjmsAdminException e) {
			e.printStackTrace();
		}
		
	}
	public static void main(String[] args) {
		new TibcoEMSMonitorTest().start();
		
	}
	
	
	

	
	
}


src/main/java/com/appdynamics/extensions/tibcoems/JavaExtensionHelper.java
package com.appdynamics.extensions.tibcoems;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import org.apache.log4j.Logger;

import com.singularity.ee.agent.systemagent.api.AManagedMonitor;
import com.singularity.ee.agent.systemagent.api.EnvPropertyWriter;
import com.singularity.ee.agent.systemagent.api.MetricWriter;
import com.singularity.ee.agent.systemagent.api.TaskExecutionContext;
import com.singularity.ee.agent.systemagent.api.TaskOutput;
import com.singularity.ee.agent.systemagent.api.exception.TaskExecutionException;

public abstract class JavaExtensionHelper extends AManagedMonitor {
	public static final Logger logger = Logger.getLogger("com.singularity.extensions.TibcoEMSMonitor");

	protected volatile String host;
	protected volatile String port;
	protected volatile String userName;
	protected volatile String passwd;
	protected volatile String restartAllowed = "TRUE";

	protected volatile Map<String, String> oldValueMap;
	protected volatile Map<String, String> valueMap;
	protected volatile long oldTime = 0;
	protected volatile long currentTime = 0;

	public abstract TaskOutput execute(Map<String, String> taskArguments,
			TaskExecutionContext taskContext) throws TaskExecutionException;

	protected void parseArgs(Map<String, String> args) {
		host = getArg(args, "host", "localhost");
		userName = getArg(args, "user", userName);
		passwd = getArg(args, "password", passwd);
		port = getArg(args, "port", "90");
		restartAllowed = getArg(args, "restartAllowed", restartAllowed);
	}

	// safe way to get parameter from monitor, but if null, use default
	protected String getArg(Map<String, String> args, String arg, String oldVal) {
		String result = args.get(arg);

		if (result == null)
			return oldVal;

		return result;
	}

	protected void printStringMetric(String name, String value) {
		String metricName = getMetricPrefix() + name;
		EnvPropertyWriter writer = this.getPropertyWriter();
		writer.printEnvironmentProperty(metricName, value);

		// just for debug output

		if (logger.isDebugEnabled()) {
			logger.debug("METRIC STRING:  NAME:" + name + " VALUE:" + value);
		}
	}

	protected void printMetric(String name, String value, String aggType,
			String timeRollup, String clusterRollup) {
		String metricName = getMetricPrefix() + name;
		MetricWriter metricWriter = getMetricWriter(metricName, aggType,
				timeRollup, clusterRollup);
		metricWriter.printMetric(value);

		// just for debug output
		if (logger.isDebugEnabled()) {
			logger.debug("METRIC:  NAME:" + metricName + " VALUE:" + value
					+ " :" + aggType + ":" + timeRollup + ":" + clusterRollup);
		}
	}

	protected String getMetricPrefix() {
		return "";
	}

	protected void startExecute(Map<String, String> taskArguments,
			TaskExecutionContext taskContext) {
		valueMap = Collections.synchronizedMap(new HashMap<String, String>());
		parseArgs(taskArguments);
	}

	protected TaskOutput finishExecute() {
		oldValueMap = valueMap;
		oldTime = currentTime;

		// just for debug output
		logger.debug("Finished METRIC COLLECTION for Monitor.......");

		return new TaskOutput("Success");
	}

	protected void close(ResultSet rs, Statement stmt, Connection conn) {
		if (rs != null) {
			try {
				rs.close();
			} catch (Exception e) {
				// ignore
			}
		}

		if (stmt != null) {
			try {
				stmt.close();
			} catch (Exception e) {
				// ignore
			}
		}

		if (conn != null) {
			try {
				conn.close();
			} catch (Exception e) {
				// ignore
			}
		}
	}

	protected String getString(float num) {
		int result = Math.round(num);
		return Integer.toString(result);
	}

	// lookup value for key, convert to float, round up or down and then return
	// as string form of int
	protected String getString(String key) {
		return getString(key, true);
	}

	// specify whether to convert this key to uppercase before looking up the
	// value
	protected String getString(String key, boolean convertUpper) {
		if (convertUpper)
			key = key.toUpperCase();

		String strResult = valueMap.get(key);

		if (strResult == null)
			return "0";

		// round the result to a integer since we don't handle fractions
		float result = Float.valueOf(strResult);
		String resultStr = getString(result);
		return resultStr;
	}

	protected String getPercent(String numerator, String denumerator) {
		float tmpResult = 0;

		try {
			tmpResult = getValue(numerator) / getValue(denumerator);
		} catch (Exception e) {
			return null;
		}
		tmpResult = tmpResult * 100;
		return getString(tmpResult);
	}

	protected String getReversePercent(float numerator, float denumerator) {
		if (denumerator == 0)
			return null;

		float tmpResult = numerator / denumerator;
		tmpResult = 1 - tmpResult;
		tmpResult = tmpResult * 100;
		return getString(tmpResult);
	}

	protected String getPercent(float numerator, float denumerator) {
		if (denumerator == 0)
			return getString(0);

		float tmpResult = numerator / denumerator;
		tmpResult = tmpResult * 100;
		return getString(tmpResult);
	}

	// math utility
	protected float getValue(String key) {
		String strResult = valueMap.get(key.toUpperCase());

		if (strResult == null)
			return 0;

		float result = Float.valueOf(strResult);
		return result;
	}

	protected float getDiffValue(String key) {
		String strResult = valueMap.get(key.toUpperCase());

		if (strResult == null)
			return 0;

		float result = Float.valueOf(strResult);
		float oldResult = 0;

		String oldResultStr = oldValueMap.get(key.toUpperCase());
		if (oldResultStr != null)
			oldResult = Float.valueOf(oldResultStr);

		return (result - oldResult);
	}

}


src/main/java/com/appdynamics/extensions/tibcoems/TibcoEMSMonitor.java
package com.appdynamics.extensions.tibcoems;

import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import com.singularity.ee.agent.systemagent.api.MetricWriter;
import com.singularity.ee.agent.systemagent.api.TaskExecutionContext;
import com.singularity.ee.agent.systemagent.api.TaskOutput;
import com.singularity.ee.agent.systemagent.api.exception.TaskExecutionException;
import com.singularity.ee.util.clock.ClockUtils;
import com.tibco.tibjms.admin.QueueInfo;
import com.tibco.tibjms.admin.ServerInfo;
import com.tibco.tibjms.admin.StatData;
import com.tibco.tibjms.admin.TibjmsAdmin;

/**
 * 
 * @author Appdynamics
 * 
 */
public class TibcoEMSMonitor extends JavaExtensionHelper {

	private volatile String tierName;
	private volatile String serverName;
	private volatile int refreshIntervalInExecutionTicks;
	private volatile int currentNumExecutionTicks = -1;
	private volatile String userid;
	private volatile String password;
	private volatile String hostname;
	private volatile String port;

	public TibcoEMSMonitor() {
		String msg = "Using Monitor Version ["+getImplementationVersion()+"]";
		logger.info(msg);
		System.out.println(msg);
		oldValueMap = Collections
				.synchronizedMap(new HashMap<String, String>());
	}

	protected void parseArgs(Map<String, String> args) {
		super.parseArgs(args);
		// if no tier defined then create metrics for all tiers
		tierName = getArg(args, "tier", null);
		userid = getArg(args, "userid", null);
		password = getArg(args, "password", null);
		hostname = getArg(args, "hostname", null);
		port = getArg(args, "port", "7222");
		serverName = getArg(args, "emsservername", null);

		int refreshIntervalSecs = Integer.parseInt(getArg(args,
				"refresh-interval", "60"));

		if (refreshIntervalSecs <= 60) {
			refreshIntervalInExecutionTicks = 1;
		} else {
			// Convert refresh interval to milliseconds and round up to the
			// nearest minute timeslice.
			// From that we can get the number of 60 second ticks before the
			// next refresh.
			// We do this to prevent time drift issues from preventing this task
			// from running.
			refreshIntervalInExecutionTicks = (int) (ClockUtils
					.roundUpTimestampToNextMinute(refreshIntervalSecs * 1000) / 60000);
		}

		if (currentNumExecutionTicks == -1) {
			// This is the first time we've parsed the args. Assume we refresh
			// the data
			// the next time we execute the monitor.
			currentNumExecutionTicks = refreshIntervalInExecutionTicks;
		}
	}

	private TibjmsAdmin connect(){
		TibjmsAdmin tibcoAdmin = null;
		try{
			tibcoAdmin = new TibjmsAdmin("tcp://" + hostname + ":"
					+ port, userid, password);
			//need improvement
		}catch(Throwable e){
			logger.debug("Issue while connecting with EMS server...",e);
		}
		return tibcoAdmin;
	}

	// collects all monitoring data for this time period from database
	private Map<String, String> putValuesIntoMap() throws Exception {
		logger.debug("Started adding metrics");
		Map<String, String> columnName2Value = new HashMap<String, String>();
		TibjmsAdmin conn = null;
		boolean debug = logger.isDebugEnabled();
		try {
			conn = connect();
			if(conn == null){
				logger.debug("Connection Failed! " + conn);
			}
			ServerInfo serverInfo = conn.getInfo();
			columnName2Value.put("ConnectionCount",
					new Integer(serverInfo.getConnectionCount()-1).toString());
			columnName2Value.put("MaxConnections",
					new Integer(serverInfo.getMaxConnections()).toString());
			columnName2Value.put("ProducerCount",
					new Long(serverInfo.getProducerCount()).toString());
			columnName2Value.put("ConsumerCount",
					new Long(serverInfo.getConsumerCount()).toString());
			columnName2Value.put("PendingMessageCount",
					new Long(serverInfo.getPendingMessageCount()).toString());
			columnName2Value.put("PendingMessageSize",
					new Long(serverInfo.getPendingMessageSize()).toString());

			columnName2Value.put("InboundMessageCount",
					new Long(serverInfo.getInboundMessageCount()).toString());
			columnName2Value.put("InboundMessageRate",
					new Long(serverInfo.getInboundMessageRate()).toString());
			columnName2Value.put("InboundByteRate",
					new Long(serverInfo.getInboundBytesRate()).toString());

			columnName2Value.put("OutboundMessageCount",
					new Long(serverInfo.getOutboundMessageCount()).toString());
			columnName2Value.put("OutboundMessageRate",
					new Long(serverInfo.getOutboundMessageRate()).toString());
			columnName2Value.put("OutboundByteRate",
					new Long(serverInfo.getOutboundBytesRate()).toString());




			// get most accurate time
			currentTime = System.currentTimeMillis();
			logger.debug("Retrieving Queue Information");
			QueueInfo[] queueInformation = conn.getQueuesStatistics();
			for (int i = 0; i < queueInformation.length; i++) {
				QueueInfo queueInfo = queueInformation[i];
				if(!queueInfo.getName().contains("$TMP$")){// skip $TMP$ queues

					columnName2Value.put(queueInfo.getName() + "|ConsumerCount",
							new Integer(queueInfo.getConsumerCount()).toString());
					columnName2Value.put(queueInfo.getName() + "|ReceiverCount",
							new Long(queueInfo.getReceiverCount()).toString());
					columnName2Value.put(queueInfo.getName()+ "|DeliveredMessageCount", 
							new Long(queueInfo.getDeliveredMessageCount()).toString());
					columnName2Value.put(queueInfo.getName()+ "|PendingMessageCount", 
							new Long(queueInfo.getPendingMessageCount()).toString());
					columnName2Value.put(queueInfo.getName()+ "|InTransitMessageCount", 
							new Long(queueInfo.getInTransitMessageCount()).toString());
					columnName2Value.put(queueInfo.getName()+ "|FlowControlMaxBytes", 
							new Long(queueInfo.getFlowControlMaxBytes()).toString());
					columnName2Value.put(queueInfo.getName()+ "|PendingMessageSize", 
							new Long(queueInfo.getPendingMessageSize()).toString());
					columnName2Value.put(queueInfo.getName() + "|MaxMsgs",
							new Long(queueInfo.getMaxMsgs()).toString());
					columnName2Value.put(queueInfo.getName() + "|MaxBytes",
							new Long(queueInfo.getMaxBytes()).toString());

					StatData inboundStatData = queueInfo.getInboundStatistics();
					if(inboundStatData != null){
						columnName2Value.put(queueInfo.getName() + "|InboundByteRate",
								new Long(inboundStatData.getByteRate()).toString());
						columnName2Value.put(queueInfo.getName() + "|InboundMessageRate",
								new Long(inboundStatData.getMessageRate()).toString());
						columnName2Value.put(queueInfo.getName() + "|InboundByteCount",
								new Long(inboundStatData.getTotalBytes()).toString());
						columnName2Value.put(queueInfo.getName() + "|InboundMessageCount",
								new Long(inboundStatData.getTotalMessages()).toString());
					}

					StatData outboundStatData = queueInfo.getOutboundStatistics();
					if(outboundStatData != null){
						columnName2Value.put(queueInfo.getName() + "|OutboundByteRate",
								new Long(outboundStatData.getByteRate()).toString());
						columnName2Value.put(queueInfo.getName() + "|OutboundMessageRate",
								new Long(outboundStatData.getMessageRate()).toString());
						columnName2Value.put(queueInfo.getName() + "|OutboundByteCount",
								new Long(outboundStatData.getTotalBytes()).toString());
						columnName2Value.put(queueInfo.getName() + "|OutboundMessageCount",
								new Long(outboundStatData.getTotalMessages()).toString());
					}


					// Compare Pending Message Size against the Max Message Size.
					if (debug) {
						logger.debug("Queue Name " + queueInfo.getName());
						logger.debug("Consumer Count "
								+ queueInfo.getConsumerCount());
						logger.debug("Delivered Message Count "
								+ queueInfo.getDeliveredMessageCount());
						logger.debug("Flow Control Max Bytes "
								+ queueInfo.getFlowControlMaxBytes());
						logger.debug("Pending Message Count "
								+ queueInfo.getPendingMessageCount());
						logger.debug("FlowControlMaxBytes"
								+ queueInfo.getFlowControlMaxBytes());
						logger.debug("MaxMessages " + queueInfo.getMaxMsgs());
						logger.debug("ReceiverCount "
								+ queueInfo.getReceiverCount());
					}
				}

			}
			logger.info("Closing Connection to Server");
			conn.close();

		} catch (com.tibco.tibjms.admin.TibjmsAdminException ex) {
			logger.error("Error connecting to EMS Server" + serverName + " "
					+ port + " " + this.hostname + " " + this.password, ex);

		} catch (Exception ex) {
			logger.error("Error getting performance data from Tibco EMS.", ex);
			throw ex;

		} finally {
			logger.info("Closing connection");
			conn.close();
		}
		return Collections.synchronizedMap(columnName2Value);
	}

	public TaskOutput execute(Map<String, String> taskArguments,
			TaskExecutionContext taskContext) throws TaskExecutionException {

		startExecute(taskArguments, taskContext);

		logger.debug("Starting METRIC COLLECTION for Tibco EMS  Monitor");

		Map<String, String> map;
		try {
			map = putValuesIntoMap();
			Iterator<String> keys = map.keySet().iterator();
			while (keys.hasNext()) {
				String key = (String) keys.next();
				String value = map.get(key);
				printMetric(key, value);
				if (logger.isDebugEnabled()) {
					logger.debug("Key :" + key + " : " + value);
				}

			}
		} catch (Exception e) {
			logger.error("Error collectng EMS metrics ", e);
		}

		return this.finishExecute();
	}

	private void printMetric(String name, String value) {
		printMetric(name, value,
				MetricWriter.METRIC_AGGREGATION_TYPE_OBSERVATION,
				MetricWriter.METRIC_TIME_ROLLUP_TYPE_CURRENT,
				MetricWriter.METRIC_CLUSTER_ROLLUP_TYPE_COLLECTIVE);

	}

	protected String getMetricPrefix() {
		if (tierName != null) {
			return "Server|Component:" + tierName + "|" + serverName + "|";
		} else {
			return "Custom Metrics|Tibco EMS Server|";
		}
	}
	public static String getImplementationVersion(){
		return TibcoEMSMonitor.class.getPackage().getImplementationTitle();
	}
}
